File: tests/csharp/Cosmos.cs
using Aspire.Microsoft.Azure.Cosmos;
using Microsoft.Azure.Cosmos;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;
using Microsoft.Extensions.Hosting;
using System;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace common.tests;

internal static class CosmosModule
{
    public static void ConfigureDatabase(IHostApplicationBuilder builder)
    {
        ConfigureCosmosClient(builder);
        builder.Services.TryAddSingleton(GetDatabase);
    }

    private static Database GetDatabase(IServiceProvider provider)
    {
        var client = provider.GetRequiredService<CosmosClient>();
        var configuration = provider.GetRequiredService<IConfiguration>();

        var databaseName = configuration.GetValue("CSHARP_COSMOS_DATABASE_NAME")
                                        .IfNone(() => configuration.GetRequiredValue("COSMOS_DATABASE_NAME"));

        return client.GetDatabase(databaseName);
    }

    private static void ConfigureCosmosClient(IHostApplicationBuilder builder)
    {
        var connectionName = builder.Configuration
                                    .GetValue("COSMOS_CONNECTION_NAME")
                                    .IfNone(string.Empty);

        builder.AddAzureCosmosClient(connectionName, configureSettings, configureClientOptions);

        void configureSettings(MicrosoftAzureCosmosSettings settings) =>
            builder.Configuration
                   .GetValue("COSMOS_CONNECTION_STRING")
                   .Iter(connectionString => settings.ConnectionString = connectionString);

        void configureClientOptions(CosmosClientOptions options)
        {
            options.EnableContentResponseOnWrite = false;
            options.UseSystemTextJsonSerializerWithOptions = JsonSerializerOptions.Web;
            options.CosmosClientTelemetryOptions = new CosmosClientTelemetryOptions { DisableDistributedTracing = false };
        }
    }
}

File: tests/fsharp/Cosmos.fs
namespace common.tests

open Aspire.Microsoft.Azure.Cosmos
open Microsoft.Azure.Cosmos
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.Hosting
open Microsoft.Extensions.DependencyInjection
open System
open System.Text.Json

[<RequireQualifiedAccess>]
module private Cosmos =
    let private configureCosmosSettings configuration (settings: MicrosoftAzureCosmosSettings) =
        configuration
        |> Configuration.getValue "COSMOS_CONNECTION_STRING"
        |> Option.iter (fun connectionString -> settings.ConnectionString <- connectionString)

    let private configureClientOptions (options: CosmosClientOptions) =
        options.EnableContentResponseOnWrite <- false
        options.UseSystemTextJsonSerializerWithOptions <- JsonSerializerOptions.Web

        options.CosmosClientTelemetryOptions <-
            let options = CosmosClientTelemetryOptions()
            options.DisableDistributedTracing <- false
            options

    let private configureClient (builder: IHostApplicationBuilder) =
        let connectionName =
            builder.Configuration
            |> Configuration.getValue "COSMOS_CONNECTION_NAME"
            |> Option.defaultValue String.Empty

        builder.AddAzureCosmosClient(
            connectionName,
            (configureCosmosSettings builder.Configuration),
            configureClientOptions
        )

        builder

    let private getDatabase (provider: IServiceProvider) =
        let client = provider.GetRequiredService<CosmosClient>()
        let configuration = provider.GetRequiredService<IConfiguration>()

        let databaseName =
            configuration
            |> Configuration.getValue "FSHARP_COSMOS_DATABASE_NAME"
            |> Option.defaultWith (fun () -> Configuration.getRequiredValue "COSMOS_DATABASE_NAME" configuration)

        client.GetDatabase(databaseName)

    let configureDatabase builder =
        builder |> configureClient |> HostApplicationBuilder.tryAddSingleton getDatabase

    let createContainer (containerName: string) (partitionKeyPath: string) (database: Database) =
        async {
            let! cancellationToken = Async.CancellationToken
            let containerProperties = ContainerProperties(containerName, partitionKeyPath)

            let! response =
                database.CreateContainerIfNotExistsAsync(containerProperties, cancellationToken = cancellationToken)
                |> Async.AwaitTask

            return response.Container
        }
